packages/vscode-extension/src/project/deviceSession.ts

```diff
@@ -102,30 +103,26 @@ export class DeviceSession implements Disposable {
 
     Logger.debug("Will wait for app ready and for preview");
     this.eventDelegate.onStateChange(StartupMessage.WaitingForAppToLoad);
-    await Promise.all([waitForAppReady, this.device.startPreview()]);
+    const [previewUrl] = await Promise.all([this.device.startPreview(), waitForAppReady]);
+    this.eventDelegate.onPreviewReady(previewUrl);
     Logger.debug("App and preview ready, moving on...");
     this.eventDelegate.onStateChange(StartupMessage.AttachingDebugger);
     await this.startDebugger();
   }
 
   public async restart() {
-    return this.launch();
+    await this.perform("restartProcess");
   }
 
-  public async start(deviceSettings: DeviceSettings, previewReadyCallback: PreviewReadyCallback) {
+  public async start(deviceSettings: DeviceSettings) {
     this.eventDelegate.onStateChange(StartupMessage.BootingDevice);
     await this.device.bootDevice();
     await this.device.changeSettings(deviceSettings);
     this.eventDelegate.onStateChange(StartupMessage.Building);
     this.buildResult = await this.disposableBuild.build;
     this.eventDelegate.onStateChange(StartupMessage.Installing);
     await this.device.installApp(this.buildResult, false);
-
-    this.device.startPreview().then(() => {
-      previewReadyCallback(this.device.previewURL!);
-    });
-
-    await this.launch();
+    await this.perform("restartProcess");
```

Comment:
not a huge fan of the perform method, it seems like it takes string and just calls back to the right method, so why can't we just call that method directly (in this case its launchApp)

Reply:
It's a good idea. I did it because it better resembled current structure for me while refactoring.